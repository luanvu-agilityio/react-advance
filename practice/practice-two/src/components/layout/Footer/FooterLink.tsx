import * as Accordion from '@radix-ui/react-accordion'
import { Box, Container, Flex } from '@radix-ui/themes'
import styled from 'styled-components'
import Link from '@components/common/Link'
import { footerSections } from '@dummy-data/footer-link'
import { ChevronDown } from 'lucide-react'
import type { MouseEvent } from 'react'
import Text from '@components/common/Text'

const StyledContainer = styled(Container)`
  width: 100%;
  width: 1260px;
  margin: 0 auto;
  background-color: transparent;
  padding: 24px 16px 32px;

  @media (min-width: 768px) {
    padding: 32px 45px 48px;
  }
`

const MobileFooter = styled(Box)`
  display: block;

  @media (min-width: 768px) {
    display: none;
  }
`

const DesktopFooter = styled(Box)`
  display: none;

  @media (min-width: 768px) {
    display: block;
  }
`

const StyledAccordionRoot = styled(Accordion.Root)`
  width: 100%;

  @media (max-width: 767px) {
    margin: 0 -16px;
  }
`
const StyledAccordionItem = styled(Accordion.Item)`
  border-bottom: 1px solid var(--black-shade-5);

  @media (max-width: 767px) {
    background: white;
    margin-left: 16px;
    border-radius: 12px;

    &:first-child {
      border-top: none;
    }
  }
`

const StyledAccordionTrigger = styled(Accordion.Trigger)`
  width: 100%;
  border: none;
  background: none;
  padding-left: 16px;
  cursor: pointer;
  transition: background-color 0.2s ease;

  @media (max-width: 767px) {
    &:hover {
      background-color: var(--black-shade-6);
    }

    &[data-state='open'] {
      background-color: var(--black-shade-6);
    }
  }

  svg {
    transition: transform 0.2s ease;
  }

  &[data-state='open'] svg {
    transform: rotate(180deg);
  }
`

const StyledAccordionContent = styled(Accordion.Content)`
  overflow: hidden;

  @media (max-width: 767px) {
    background-color: var(--black-shade-6);
    padding: 0 16px;

    &[data-state='open'] {
      animation: slideDown 300ms cubic-bezier(0.87, 0, 0.13, 1);
    }

    &[data-state='closed'] {
      animation: slideUp 300ms cubic-bezier(0.87, 0, 0.13, 1);
    }
  }

  @keyframes slideDown {
    from {
      height: 0;
      opacity: 0;
    }
    to {
      height: var(--radix-accordion-content-height);
      opacity: 1;
    }
  }

  @keyframes slideUp {
    from {
      height: var(--radix-accordion-content-height);
      opacity: 1;
    }
    to {
      height: 0;
      opacity: 0;
    }
  }
`

const AccordionHeader = styled(Flex)`
  @media (max-width: 767px) {
    min-height: 48px;
  }
`

const LinkList = styled.ul`
  list-style: none;
  padding: 0;
  margin: 0;

  @media (max-width: 767px) {
    padding: 8px 0 16px;
  }
`

const LinkItem = styled.li`
  margin-bottom: 16px;
  font-family: var(--font-family-secondary);
  font-size: 13px;
  font-weight: var(--font-weight-regular);

  @media (max-width: 767px) {
    margin-bottom: 12px;

    a {
      display: block;
      padding: 8px 0;
      color: var(--green-color-default);

      &:active {
        opacity: 0.7;
      }
    }
  }

  &:last-child {
    margin-bottom: 0;
  }
`

const FooterLink = () => {
  const handleLinkClick = (event: MouseEvent<HTMLAnchorElement>) => {
    console.log('Link clicked:', event.currentTarget.href)
  }

  return (
    <StyledContainer>
      <MobileFooter>
        <StyledAccordionRoot type="multiple">
          {footerSections.map((section) => (
            <StyledAccordionItem key={section.title} value={section.title}>
              <StyledAccordionTrigger>
                <AccordionHeader justify="between" align="center">
                  <Text
                    text={section.title}
                    as="h3"
                    style={{
                      color: 'var(--black-color-default)',
                      fontSize: '18px',
                      fontWeight: 'var(--font-weight-semibold)',
                      margin: 0,
                    }}
                  />
                  <ChevronDown
                    size={20}
                    color="var(--black-color-default)"
                    style={{ opacity: 0.7 }}
                  />
                </AccordionHeader>
              </StyledAccordionTrigger>
              <StyledAccordionContent>
                <LinkList>
                  {section.links.map((link) => (
                    <LinkItem key={link.text}>
                      <Link
                        href={link.href}
                        onClick={handleLinkClick}
                        target={
                          link.href.startsWith('http') ? '_blank' : '_self'
                        }
                      >
                        {link.text}
                      </Link>
                    </LinkItem>
                  ))}
                </LinkList>
              </StyledAccordionContent>
            </StyledAccordionItem>
          ))}
        </StyledAccordionRoot>
      </MobileFooter>

      {/* Desktop Footer with Grid */}
      <DesktopFooter>
        <Flex gap="6" wrap="wrap">
          {footerSections.map((section) => (
            <Box
              key={section.title}
              style={{
                flex:
                  footerSections[footerSections.length - 1].title ===
                  section.title
                    ? 'none'
                    : '1 1 200px',
              }}
            >
              <Text
                text={section.title}
                as="h3"
                style={{
                  fontSize: '18px',
                  fontWeight: 'var(--font-weight-semibold)',
                  marginBottom: '12px',
                }}
              />
              <LinkList>
                {section.links.map((link) => (
                  <LinkItem key={link.text}>
                    <Link
                      href={link.href}
                      onClick={handleLinkClick}
                      target={link.href.startsWith('http') ? '_blank' : '_self'}
                    >
                      {link.text}
                    </Link>
                  </LinkItem>
                ))}
              </LinkList>
            </Box>
          ))}
        </Flex>
      </DesktopFooter>
    </StyledContainer>
  )
}

export default FooterLink
