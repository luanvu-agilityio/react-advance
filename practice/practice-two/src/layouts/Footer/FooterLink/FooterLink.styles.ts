import styled from 'styled-components'
import * as Accordion from '@radix-ui/react-accordion'
import { Box, Container, Flex } from '@radix-ui/themes'

export const StyledContainer = styled(Container)`
  width: 100%;
  width: 1260px;
  margin: 0 auto;
  background-color: transparent;
  padding: 24px 16px 32px;

  @media (min-width: 768px) {
    padding: 32px 45px 48px;
  }
`

export const MobileFooter = styled(Box)`
  display: block;

  @media (min-width: 768px) {
    display: none;
  }
`

export const DesktopFooter = styled(Box)`
  display: none;

  @media (min-width: 768px) {
    display: block;
  }
`

export const StyledAccordionRoot = styled(Accordion.Root)`
  width: 100%;

  @media (max-width: 767px) {
    margin: 0 -16px;
  }
`
export const StyledAccordionItem = styled(Accordion.Item)`
  border-bottom: 1px solid var(--black-shade-5);

  @media (max-width: 767px) {
    background: white;
    margin-left: 16px;
    border-radius: 12px;

    &:first-child {
      border-top: none;
    }
  }
`

export const StyledAccordionTrigger = styled(Accordion.Trigger)`
  width: 100%;
  border: none;
  background: none;
  padding-left: 16px;
  cursor: pointer;
  transition: background-color 0.2s ease;

  @media (max-width: 767px) {
    &:hover {
      background-color: var(--black-shade-6);
    }

    &[data-state='open'] {
      background-color: var(--black-shade-6);
    }
  }

  svg {
    transition: transform 0.2s ease;
  }

  &[data-state='open'] svg {
    transform: rotate(180deg);
  }
`

export const StyledAccordionContent = styled(Accordion.Content)`
  overflow: hidden;

  @media (max-width: 767px) {
    background-color: var(--black-shade-6);
    padding: 0 16px;

    &[data-state='open'] {
      animation: slideDown 300ms cubic-bezier(0.87, 0, 0.13, 1);
    }

    &[data-state='closed'] {
      animation: slideUp 300ms cubic-bezier(0.87, 0, 0.13, 1);
    }
  }

  @keyframes slideDown {
    from {
      height: 0;
      opacity: 0;
    }
    to {
      height: var(--radix-accordion-content-height);
      opacity: 1;
    }
  }

  @keyframes slideUp {
    from {
      height: var(--radix-accordion-content-height);
      opacity: 1;
    }
    to {
      height: 0;
      opacity: 0;
    }
  }
`

export const AccordionHeader = styled(Flex)`
  @media (max-width: 767px) {
    min-height: 48px;
  }
`

export const LinkList = styled.ul`
  list-style: none;
  padding: 0;
  margin: 0;

  @media (max-width: 767px) {
    padding: 8px 0 16px;
  }
`

export const LinkItem = styled.li`
  margin-bottom: 16px;
  font-family: var(--font-family-secondary);
  font-size: 13px;
  font-weight: var(--font-weight-regular);

  @media (max-width: 767px) {
    margin-bottom: 12px;

    a {
      display: block;
      padding: 8px 0;
      color: var(--green-color-default);

      &:active {
        opacity: 0.7;
      }
    }
  }

  &:last-child {
    margin-bottom: 0;
  }
`
