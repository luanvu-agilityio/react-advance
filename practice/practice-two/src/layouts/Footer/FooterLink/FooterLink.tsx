import { Box, Flex } from '@radix-ui/themes'
import Link from '@components/common/Link'
import { footerSections } from '@constants/footer-link'
import { ChevronDown } from 'lucide-react'
import { memo, type MouseEvent } from 'react'
import Text from '@components/common/Text/index'
import {
  AccordionHeader,
  DesktopFooter,
  LinkItem,
  LinkList,
  MobileFooter,
  StyledAccordionContent,
  StyledAccordionItem,
  StyledAccordionRoot,
  StyledAccordionTrigger,
  StyledContainer,
} from './FooterLink.styles'

interface FooterLinkProps {
  sections?: {
    title: string
    links: {
      text: string
      href: string
    }[]
  }[]
}

const FooterLink = memo(({ sections = footerSections }: FooterLinkProps) => {
  const handleLinkClick = (event: MouseEvent<HTMLAnchorElement>) => {
    event.preventDefault?.()
  }

  return (
    <StyledContainer>
      <MobileFooter>
        <StyledAccordionRoot type="multiple">
          {sections.map((section) => (
            <StyledAccordionItem key={section.title} value={section.title}>
              <StyledAccordionTrigger>
                <AccordionHeader justify="between" align="center">
                  <Text
                    text={section.title}
                    as="h3"
                    style={{
                      color: 'var(--black-color-default)',
                      fontSize: '18px',
                      fontWeight: 'var(--font-weight-semibold)',
                      margin: 0,
                    }}
                  />
                  <ChevronDown
                    size={20}
                    color="var(--black-color-default)"
                    style={{ opacity: 0.7 }}
                  />
                </AccordionHeader>
              </StyledAccordionTrigger>
              <StyledAccordionContent>
                <LinkList>
                  {section.links.map((link) => (
                    <LinkItem key={link.text}>
                      <Link href={link.href} onClick={handleLinkClick}>
                        {link.text}
                      </Link>
                    </LinkItem>
                  ))}
                </LinkList>
              </StyledAccordionContent>
            </StyledAccordionItem>
          ))}
        </StyledAccordionRoot>
      </MobileFooter>

      {/* Desktop Footer with Grid */}
      <DesktopFooter>
        <Flex gap="6" wrap="wrap">
          {sections.map((section) => (
            <Box
              key={section.title}
              style={{
                flex:
                  sections[sections.length - 1].title === section.title
                    ? 'none'
                    : '1 1 200px',
              }}
            >
              <Text
                text={section.title}
                as="h3"
                style={{
                  fontSize: '18px',
                  fontWeight: 'var(--font-weight-semibold)',
                  marginBottom: '12px',
                }}
              />
              <LinkList>
                {section.links.map((link) => (
                  <LinkItem key={link.text}>
                    <Link href={link.href} onClick={handleLinkClick}>
                      {link.text}
                    </Link>
                  </LinkItem>
                ))}
              </LinkList>
            </Box>
          ))}
        </Flex>
      </DesktopFooter>
    </StyledContainer>
  )
})

export default FooterLink
