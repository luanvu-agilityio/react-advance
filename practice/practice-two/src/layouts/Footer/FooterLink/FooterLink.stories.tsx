import type { Meta, StoryObj } from '@storybook/react'
import { Theme as ThemeProvider } from '@radix-ui/themes'
import FooterLink from './FooterLink'

// Mock the footer sections data
const mockFooterSections = [
  {
    title: 'About',
    links: [
      { text: 'About Us', href: '/about' },
      { text: 'Careers', href: '/careers' },
      { text: 'Press', href: '/press' },
      { text: 'Blog', href: '/blog' },
    ],
  },
  {
    title: 'Customer Service',
    links: [
      { text: 'Contact Us', href: '/contact' },
      { text: 'Help Center', href: '/help' },
      { text: 'Shipping Info', href: '/shipping' },
      { text: 'Returns', href: '/returns' },
    ],
  },
  {
    title: 'Legal',
    links: [
      { text: 'Privacy Policy', href: '/privacy' },
      { text: 'Terms of Service', href: '/terms' },
      { text: 'Cookie Policy', href: '/cookies' },
    ],
  },
  {
    title: 'Connect',
    links: [
      { text: 'Facebook', href: 'https://facebook.com' },
      { text: 'Twitter', href: 'https://twitter.com' },
      { text: 'Instagram', href: 'https://instagram.com' },
      { text: 'LinkedIn', href: 'https://linkedin.com' },
    ],
  },
]

const meta: Meta<typeof FooterLink> = {
  title: 'Layout/Footer/FooterLink',
  component: FooterLink,
  parameters: {
    layout: 'fullscreen',
    docs: {
      description: {
        component:
          'Responsive footer component that displays as an accordion on mobile and grid layout on desktop.',
      },
    },
    viewport: {
      defaultViewport: 'responsive',
    },
  },
  decorators: [
    (Story) => (
      <ThemeProvider>
        <div
          style={{
            backgroundColor: '#f5f5f5',
            minHeight: '100vh',
            padding: '20px 0',
          }}
        >
          <Story />
        </div>
      </ThemeProvider>
    ),
  ],
}

export default meta
type Story = StoryObj<typeof FooterLink>

export const Desktop: Story = {
  args: {
    sections: mockFooterSections,
  },
  parameters: {
    viewport: {
      defaultViewport: 'desktop',
    },
    docs: {
      description: {
        story: 'Desktop view showing the footer links in a grid layout',
      },
    },
  },
}

export const Tablet: Story = {
  parameters: {
    viewport: {
      defaultViewport: 'tablet',
    },
    docs: {
      description: {
        story: 'Tablet view showing the footer links in a grid layout',
      },
    },
  },
}

export const Mobile: Story = {
  parameters: {
    viewport: {
      defaultViewport: 'mobile1',
    },
    docs: {
      description: {
        story: 'Mobile view showing the footer links as collapsible accordions',
      },
    },
  },
}

export const ExternalLinks: Story = {
  decorators: [
    () => {
      return (
        <ThemeProvider>
          <div
            style={{
              backgroundColor: '#f5f5f5',
              minHeight: '100vh',
              padding: '20px 0',
            }}
          >
            <FooterLink
              sections={[
                {
                  title: 'External Partners',
                  links: [
                    { text: 'Google', href: 'https://google.com' },
                    { text: 'Facebook', href: 'https://facebook.com' },
                    { text: 'Amazon', href: 'https://amazon.com' },
                    { text: 'Microsoft', href: 'https://microsoft.com' },
                  ],
                },
              ]}
            />
          </div>
        </ThemeProvider>
      )
    },
  ],
  parameters: {
    docs: {
      description: {
        story: 'Footer component with external links that open in new tabs',
      },
    },
  },
}
