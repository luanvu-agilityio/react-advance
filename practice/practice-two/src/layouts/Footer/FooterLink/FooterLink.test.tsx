<<<<<<< HEAD
import { render, screen } from '@testing-library/react'
import { Theme as ThemeProvider } from '@radix-ui/themes'
import type { ReactElement } from 'react'

// Mock the footer sections BEFORE importing the component
jest.mock('@constants/footer-link', () => ({
  footerSections: [
    {
      title: 'Get in touch',
      links: [
        { text: 'About Us', href: '/about' },
        { text: 'Careers', href: '/careers' },
      ],
    },
    {
      title: 'Connections',
      links: [
        { text: 'Facebook', href: 'https://facebook.com' },
        { text: 'Twitter', href: 'https://twitter.com' },
      ],
    },
    {
      title: 'Earnings',
      links: [
        { text: 'Become an Affiliate', href: '/affiliate' },
        { text: 'Sell on Market', href: '/sell' },
      ],
    },
  ],
}))

import FooterLink from './FooterLink'

// Helper to render with ThemeProvider
const renderWithTheme = (component: ReactElement) => {
  return render(<ThemeProvider>{component}</ThemeProvider>)
}

describe('FooterLink Component', () => {
  afterAll(() => {
    jest.restoreAllMocks()
  })

  // SNAPSHOT TESTS
  describe('Snapshot Tests', () => {
    it('renders correctly', () => {
      const { container } = renderWithTheme(<FooterLink />)
      expect(container).toMatchSnapshot()
    })
  })

  // BASIC RENDERING TESTS
  describe('Basic Rendering', () => {
    it('renders all footer links', () => {
      renderWithTheme(<FooterLink />)
      expect(screen.getByText('About Us')).toBeInTheDocument()
      expect(screen.getByText('Facebook')).toBeInTheDocument()
      expect(screen.getByText('Become an Affiliate')).toBeInTheDocument()
    })

    it('applies correct attributes to external links', () => {
      renderWithTheme(<FooterLink />)
      const facebookLink = screen.getByText('Facebook').closest('a')
      expect(facebookLink).toHaveAttribute('target', '_blank')
      expect(facebookLink).toHaveAttribute('rel', 'noopener noreferrer')
    })

    it('applies correct attributes to internal links', () => {
      renderWithTheme(<FooterLink />)
      const aboutLink = screen.getByText('About Us').closest('a')
      expect(aboutLink).toHaveAttribute('target', '_self')
      expect(aboutLink).not.toHaveAttribute('rel')
    })
  })

  // CUSTOM SECTIONS TEST
  describe('Custom Sections', () => {
    it('renders custom sections when provided', () => {
      const customSections = [
        {
          title: 'Custom Section',
          links: [
            { text: 'Custom Link 1', href: '/custom1' },
            { text: 'Custom Link 2', href: '/custom2' },
          ],
        },
      ]

      renderWithTheme(<FooterLink sections={customSections} />)

      expect(screen.getByText('Custom Link 1')).toBeInTheDocument()
      expect(screen.getByText('Custom Link 2')).toBeInTheDocument()
      expect(screen.queryByText('Get in touch')).not.toBeInTheDocument()
    })

    it('handles empty sections gracefully', () => {
      const { container } = renderWithTheme(<FooterLink sections={[]} />)
      expect(container).toBeInTheDocument()
      expect(screen.queryByText('Get in touch')).not.toBeInTheDocument()
=======
import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import { Theme as ThemeProvider } from '@radix-ui/themes'
import FooterLink from './FooterLink'

// Mock the footer sections data
const mockFooterSections = [
  {
    title: 'About',
    links: [
      { text: 'About Us', href: '/about' },
      { text: 'Careers', href: '/careers' },
      { text: 'Press', href: '/press' },
    ],
  },
  {
    title: 'Customer Service',
    links: [
      { text: 'Contact Us', href: '/contact' },
      { text: 'Help Center', href: '/help' },
    ],
  },
  {
    title: 'Connect',
    links: [
      { text: 'Facebook', href: 'https://facebook.com' },
      { text: 'Twitter', href: 'https://twitter.com' },
    ],
  },
]

// Mock the footer sections constant
jest.mock('@constants/footer-link', () => ({
  footerSections: mockFooterSections,
}))

// Mock console.log to test click handlers
const mockConsoleLog = jest.spyOn(console, 'log').mockImplementation()

const renderWithTheme = (component: React.ReactElement) => {
  return render(<ThemeProvider>{component}</ThemeProvider>)
}

// Mock window.matchMedia for responsive tests
const mockMatchMedia = (matches: boolean) => {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  })
}

describe('FooterLink Component', () => {
  beforeEach(() => {
    mockConsoleLog.mockClear()
  })

  afterEach(() => {
    jest.restoreAllMocks()
  })

  it('renders all footer sections', () => {
    renderWithTheme(<FooterLink />)

    expect(screen.getAllByText('About')).toHaveLength(2) // Mobile + Desktop
    expect(screen.getAllByText('Customer Service')).toHaveLength(2)
    expect(screen.getAllByText('Connect')).toHaveLength(2)
  })

  it('renders all footer links', () => {
    renderWithTheme(<FooterLink />)

    expect(screen.getAllByText('About Us')).toHaveLength(2)
    expect(screen.getAllByText('Careers')).toHaveLength(2)
    expect(screen.getAllByText('Contact Us')).toHaveLength(2)
    expect(screen.getAllByText('Facebook')).toHaveLength(2)
  })

  describe('Mobile View', () => {
    beforeEach(() => {
      mockMatchMedia(true) // Mobile breakpoint
    })

    it('shows mobile accordion and hides desktop grid', () => {
      renderWithTheme(<FooterLink />)

      // Check that mobile accordion is present
      const accordionTriggers = screen.getAllByRole('button')
      expect(accordionTriggers.length).toBeGreaterThan(0)

      // Check accordion structure
      expect(screen.getByText('About')).toBeInTheDocument()
    })

    it('expands accordion when clicked', async () => {
      renderWithTheme(<FooterLink />)

      const aboutButton = screen.getAllByText('About')[0].closest('button')
      expect(aboutButton).toBeInTheDocument()

      fireEvent.click(aboutButton!)

      // Wait for accordion animation
      await waitFor(() => {
        expect(aboutButton).toHaveAttribute('data-state', 'open')
      })
    })

    it('rotates chevron icon when accordion is opened', async () => {
      renderWithTheme(<FooterLink />)

      const aboutButton = screen.getAllByText('About')[0].closest('button')
      fireEvent.click(aboutButton!)

      await waitFor(() => {
        const chevronIcon = aboutButton!.querySelector('svg')
        expect(chevronIcon).toHaveStyle('transform: rotate(180deg)')
      })
    })

    it('allows multiple accordions to be open simultaneously', async () => {
      renderWithTheme(<FooterLink />)

      const aboutButton = screen.getAllByText('About')[0].closest('button')
      const serviceButton = screen
        .getAllByText('Customer Service')[0]
        .closest('button')

      fireEvent.click(aboutButton!)
      fireEvent.click(serviceButton!)

      await waitFor(() => {
        expect(aboutButton).toHaveAttribute('data-state', 'open')
        expect(serviceButton).toHaveAttribute('data-state', 'open')
      })
    })
  })

  describe('Desktop View', () => {
    beforeEach(() => {
      mockMatchMedia(false) // Desktop breakpoint
    })

    it('shows desktop grid layout', () => {
      renderWithTheme(<FooterLink />)

      // Desktop version should not have accordion buttons
      const accordionButtons = screen.queryAllByRole('button')
      expect(accordionButtons).toHaveLength(0)
    })

    it('displays all links in grid format', () => {
      renderWithTheme(<FooterLink />)

      expect(screen.getByText('About Us')).toBeInTheDocument()
      expect(screen.getByText('Careers')).toBeInTheDocument()
      expect(screen.getByText('Contact Us')).toBeInTheDocument()
    })
  })

  describe('Link Functionality', () => {
    it('handles internal link clicks', () => {
      renderWithTheme(<FooterLink />)

      const aboutUsLink = screen.getAllByText('About Us')[0]
      fireEvent.click(aboutUsLink)

      expect(mockConsoleLog).toHaveBeenCalledWith('Link clicked:', '/about')
    })

    it('handles external link clicks', () => {
      renderWithTheme(<FooterLink />)

      const facebookLink = screen.getAllByText('Facebook')[0]
      fireEvent.click(facebookLink)

      expect(mockConsoleLog).toHaveBeenCalledWith(
        'Link clicked:',
        'https://facebook.com'
      )
    })

    it('sets correct target for external links', () => {
      renderWithTheme(<FooterLink />)

      const facebookLink = screen.getAllByText('Facebook')[0]
      expect(facebookLink).toHaveAttribute('target', '_blank')
    })
  })

  describe('Accessibility', () => {
    it('has proper ARIA attributes for accordions', () => {
      renderWithTheme(<FooterLink />)

      const aboutButton = screen.getAllByText('About')[0].closest('button')
      expect(aboutButton).toHaveAttribute('aria-expanded', 'false')
    })

    it('updates ARIA attributes when accordion is expanded', async () => {
      renderWithTheme(<FooterLink />)

      const aboutButton = screen.getAllByText('About')[0].closest('button')
      fireEvent.click(aboutButton!)

      await waitFor(() => {
        expect(aboutButton).toHaveAttribute('aria-expanded', 'true')
      })
    })

    it('has keyboard navigation support', () => {
      renderWithTheme(<FooterLink />)

      const aboutButton = screen.getAllByText('About')[0].closest('button')

      // Test Enter key
      fireEvent.keyDown(aboutButton!, { key: 'Enter', code: 'Enter' })
      fireEvent.keyUp(aboutButton!, { key: 'Enter', code: 'Enter' })

      // Test Space key
      fireEvent.keyDown(aboutButton!, { key: ' ', code: 'Space' })
      fireEvent.keyUp(aboutButton!, { key: ' ', code: 'Space' })
    })
  })

  describe('Edge Cases', () => {
    it('handles empty footer sections gracefully', () => {
      // Mock empty sections
      jest.doMock('@constants/footer-link', () => ({
        footerSections: [],
      }))

      const { container } = renderWithTheme(<FooterLink />)
      expect(container).toBeInTheDocument()
>>>>>>> 7a35d9791a5da6fe80ff0a8541efaf78233de04d
    })
  })
})
