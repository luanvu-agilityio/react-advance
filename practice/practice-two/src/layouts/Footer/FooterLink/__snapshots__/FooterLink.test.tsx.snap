// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FooterLink Component Snapshot Tests renders correctly 1`] = `
<div>
  <div
    class="radix-themes"
    data-accent-color="indigo"
    data-gray-color="slate"
    data-has-background="true"
    data-is-root-theme="true"
    data-panel-background="translucent"
    data-radius="medium"
    data-scaling="100%"
  >
    <div
      class="rt-Container rt-r-size-4 sc-fYsHOw dzqQTM"
    >
      <div
        class="rt-ContainerInner"
      >
        <div
          class="rt-Box sc-dKREkF ewMOXs"
        >
          <div
            class="sc-dIMoHT fiEXjO"
            data-orientation="vertical"
          >
            <div
              class="sc-iQQCXo hIsANv"
              data-orientation="vertical"
              data-state="closed"
            >
              <button
                aria-controls="radix-«r1»"
                aria-expanded="false"
                class="sc-gDpztx hFqOKu"
                data-orientation="vertical"
                data-radix-collection-item=""
                data-state="closed"
                id="radix-«r0»"
                type="button"
              >
                <div
                  class="rt-Flex rt-r-ai-center rt-r-jc-space-between sc-icnseD kgvDEc"
                >
                  <h3
                    style="font-size: 18px; font-weight: var(--font-weight-semibold); margin: 0px;"
                  >
                    Get in touch
                  </h3>
                  <svg
                    aria-hidden="true"
                    class="lucide lucide-chevron-down"
                    fill="none"
                    height="20"
                    stroke="var(--black-color-default)"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    style="opacity: 0.7;"
                    viewBox="0 0 24 24"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m6 9 6 6 6-6"
                    />
                  </svg>
                </div>
              </button>
              <div
                aria-labelledby="radix-«r0»"
                class="sc-kpOvIu jjvMs"
                data-orientation="vertical"
                data-state="closed"
                hidden=""
                id="radix-«r1»"
                role="region"
                style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
              />
            </div>
            <div
              class="sc-iQQCXo hIsANv"
              data-orientation="vertical"
              data-state="closed"
            >
              <button
                aria-controls="radix-«r3»"
                aria-expanded="false"
                class="sc-gDpztx hFqOKu"
                data-orientation="vertical"
                data-radix-collection-item=""
                data-state="closed"
                id="radix-«r2»"
                type="button"
              >
                <div
                  class="rt-Flex rt-r-ai-center rt-r-jc-space-between sc-icnseD kgvDEc"
                >
                  <h3
                    style="font-size: 18px; font-weight: var(--font-weight-semibold); margin: 0px;"
                  >
                    Connections
                  </h3>
                  <svg
                    aria-hidden="true"
                    class="lucide lucide-chevron-down"
                    fill="none"
                    height="20"
                    stroke="var(--black-color-default)"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    style="opacity: 0.7;"
                    viewBox="0 0 24 24"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m6 9 6 6 6-6"
                    />
                  </svg>
                </div>
              </button>
              <div
                aria-labelledby="radix-«r2»"
                class="sc-kpOvIu jjvMs"
                data-orientation="vertical"
                data-state="closed"
                hidden=""
                id="radix-«r3»"
                role="region"
                style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
              />
            </div>
            <div
              class="sc-iQQCXo hIsANv"
              data-orientation="vertical"
              data-state="closed"
            >
              <button
                aria-controls="radix-«r5»"
                aria-expanded="false"
                class="sc-gDpztx hFqOKu"
                data-orientation="vertical"
                data-radix-collection-item=""
                data-state="closed"
                id="radix-«r4»"
                type="button"
              >
                <div
                  class="rt-Flex rt-r-ai-center rt-r-jc-space-between sc-icnseD kgvDEc"
                >
                  <h3
                    style="font-size: 18px; font-weight: var(--font-weight-semibold); margin: 0px;"
                  >
                    Earnings
                  </h3>
                  <svg
                    aria-hidden="true"
                    class="lucide lucide-chevron-down"
                    fill="none"
                    height="20"
                    stroke="var(--black-color-default)"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    style="opacity: 0.7;"
                    viewBox="0 0 24 24"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m6 9 6 6 6-6"
                    />
                  </svg>
                </div>
              </button>
              <div
                aria-labelledby="radix-«r4»"
                class="sc-kpOvIu jjvMs"
                data-orientation="vertical"
                data-state="closed"
                hidden=""
                id="radix-«r5»"
                role="region"
                style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
              />
            </div>
          </div>
        </div>
        <div
          class="rt-Box sc-fWnslK hoHqUa"
        >
          <div
            class="rt-Flex rt-r-fw-wrap rt-r-gap-6"
          >
            <div
              class="rt-Box"
              style="flex: 1 1 200px;"
            >
              <h3
                style="font-size: 18px; font-weight: var(--font-weight-semibold); margin-bottom: 12px;"
              >
                Get in touch
              </h3>
              <ul
                class="sc-jMsorb ddihlD"
              >
                <li
                  class="sc-cXawGu twrLz"
                >
                  <a
                    class="sc-Qotzb jUyYcI"
                    href="/about"
                    target="_self"
                  >
                    About Us
                  </a>
                </li>
                <li
                  class="sc-cXawGu twrLz"
                >
                  <a
                    class="sc-Qotzb jUyYcI"
                    href="/careers"
                    target="_self"
                  >
                    Careers
                  </a>
                </li>
              </ul>
            </div>
            <div
              class="rt-Box"
              style="flex: 1 1 200px;"
            >
              <h3
                style="font-size: 18px; font-weight: var(--font-weight-semibold); margin-bottom: 12px;"
              >
                Connections
              </h3>
              <ul
                class="sc-jMsorb ddihlD"
              >
                <li
                  class="sc-cXawGu twrLz"
                >
                  <a
                    class="sc-Qotzb jUyYcI"
                    href="https://facebook.com"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    Facebook
                  </a>
                </li>
                <li
                  class="sc-cXawGu twrLz"
                >
                  <a
                    class="sc-Qotzb jUyYcI"
                    href="https://twitter.com"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    Twitter
                  </a>
                </li>
              </ul>
            </div>
            <div
              class="rt-Box"
              style="flex: 0 0 auto;"
            >
              <h3
                style="font-size: 18px; font-weight: var(--font-weight-semibold); margin-bottom: 12px;"
              >
                Earnings
              </h3>
              <ul
                class="sc-jMsorb ddihlD"
              >
                <li
                  class="sc-cXawGu twrLz"
                >
                  <a
                    class="sc-Qotzb jUyYcI"
                    href="/affiliate"
                    target="_self"
                  >
                    Become an Affiliate
                  </a>
                </li>
                <li
                  class="sc-cXawGu twrLz"
                >
                  <a
                    class="sc-Qotzb jUyYcI"
                    href="/sell"
                    target="_self"
                  >
                    Sell on Market
                  </a>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
